### PR TITLE
add iOS and armv8 build

### PR DESCRIPTION
**Pull request by vrnmthr**
### Description
Adds iOS compilation support to marian, and adds memory-mapping support to TranslateService in `translator.h`. Everything should be backwards-compatible. Builds off of the work done by @XapaJIaMnu in  https://github.com/marian-nmt/marian-dev/pull/1009. Compilation for iOS is intended to be done using the added cmake toolchain file. It is detected via `CMAKE_SYSTEM_NAME` in the cmake build. 

Also fixes one bug in the memory mapping in `binary.cpp` that allows us to read mmap files with fewer memory allocations.

Added dependencies: simd_utils, ruy

### How to test
Build `cmake . -Bbuild -DCOMPILE_CUDA=off -DCMAKE_TOOLCHAIN_FILE=./cmake/ios.toolchain.cmake -DPLATFORM=OS64 -DDEPLOYMENT_TARGET=13.0` for iOS

Build `cmake . -Bbuild -DCOMPILE_CUDA=off -DARM=on` for ARM

### Checklist

- [x] I have tested the code manually
- [ ] I have run regression tests
- [x] I have read and followed CONTRIBUTING.md
- [ ] I have updated CHANGELOG.md

### CODE REVIEW

File: src/3rd_party/spdlog/async_logger.h
Comment (hieuhoang) [inline src/3rd_party/spdlog/async_logger.h:1]: Any way

**Review by hieuhoang — APPROVED** [review]

File: .gitmodules
Comment (snukky) [inline .gitmodules:23]: For all other submodules, we have their frozen forks in marian-nmt. I think we can have forks for ruy and simd_utils as well. This prevents from breakable changes, but delays pulling recent fixes, which seems what we prefer.

File: CMakeLists.txt
Comment (snukky) [inline CMakeLists.txt:108]: Can this be problematic if ArmV8 is not available? Will there be a clear warning or error message if that's the case?

File: CMakeLists.txt
Comment (snukky) [inline CMakeLists.txt:287]: I wonder why we were using `-march` and if it's okay to drop it in all builds (probably yes).

File: src/3rd_party/CMakeLists.txt
Comment (snukky) [inline src/3rd_party/CMakeLists.txt:17]: Is this worth a short comment in the code why it's disabled?

File: src/3rd_party/spdlog/async_logger.h
Comment (snukky) [reply src/3rd_party/spdlog/async_logger.h:1]: I agree with Hieu, if this is only cosmetics, I would suggest not doing these changes. If it's required, that's fine too since we probably already modified those 3rd party files; but let's add a comment as well so that it will be clearer what changes should be kept when one updates the 3rd party package eventually.

File: src/functional/operators.h
Comment (snukky) [inline src/functional/operators.h:223]: Do we need to address the todo comment in some way before merging or not?

File: src/translator/translator.h
Comment (snukky) [inline src/translator/translator.h:303]: We should check if additional logging doesn't break some regression tests, but I think it does not.

**Review by snukky — APPROVED** [review]
LGTM, I only had minor suggestions and clarifying questions.

File: src/translator/translator.h
Comment (vrnmthr) [reply src/translator/translator.h:303]: whenever we run the regression tests we can verify this

File: .gitmodules
Comment (vrnmthr) [reply .gitmodules:23]: all changes we needed to these are already in upstream. until we find we need to make local changes, doesn't it make sense to keep upstream? they can't do breakable changes to us until we manually bump our submodule commit.

File: CMakeLists.txt
Comment (vrnmthr) [reply CMakeLists.txt:287]: probably in there for additional optimizations for people building & running marian locally. if you confirm I can take it out then I can simplify the logic a bit

File: src/3rd_party/CMakeLists.txt
Comment (vrnmthr) [reply src/3rd_party/CMakeLists.txt:17]: it is because the default codepath does not use ruy so we don't want to add the subdirectories to the build unless it is explicitly enabled. I can add a comment

File: src/3rd_party/spdlog/async_logger.h
Comment (vrnmthr) [reply src/3rd_party/spdlog/async_logger.h:1]: they are not just cosmetic, in apple xcode <> syntax means it will look in system libs. "" syntax means it will look in user libs. so without these changes I cannot import marian .h files in xcode to integrate into iOS app

Comment (vrnmthr) [reply src/3rd_party/spdlog/async_logger.h:1]: I am happy to add a comment but since there are so many changed files, where do you want me to add the comment? in every file?

Comment (hieuhoang) [reply src/3rd_party/spdlog/async_logger.h:1]: <> means default folder for any c/c++ compiler. Usually solveable by adding -I [folder with .h files] arg to compiler. That has to be done for both CMake and Xcode.

File: .gitmodules
Comment (snukky) [reply .gitmodules:23]: Okay, then that's fine. Thanks.

File: src/3rd_party/CMakeLists.txt
Comment (snukky) [reply src/3rd_party/CMakeLists.txt:17]: Adding a comment will be useful.

File: src/translator/translator.h
Comment (snukky) [reply src/translator/translator.h:303]: All regression tests pass with the following compilation options: `CC=/usr/bin/gcc-8 CXX=/usr/bin/g++-8 CUDAHOSTCXX=/usr/bin/g++8 cmake -DDETERMINISTIC=on -DUSE_SENTENCEPIECE=ON -DUSE_FBGEMM=on -DCOMPILE_CPU=on -DCOMPILE_TESTS=ON -DCOMPILE_EXAMPLES=ON -DCOMPILE_SERVER=ON -DCUDA_TOOLKIT_ROOT_DIR=/usr/local/cuda-11.1 -DCOMPILE_MAXWELL=on -DCOMPILE_PASCAL=on ..`

Is anything missing there?

Comment (vrnmthr) [reply src/translator/translator.h:303]: thanks for running those roman. that is not building for iOS, but I think that should be fine. that at least shows us that the canonical codepath has no regressions

File: src/3rd_party/CMakeLists.txt
Comment (vrnmthr) [reply src/3rd_party/CMakeLists.txt:17]: added comment

File: src/3rd_party/spdlog/async_logger.h
Comment (vrnmthr) [reply src/3rd_party/spdlog/async_logger.h:1]: I will revert these changes for now and see if we can fix at build-time

Comment (vrnmthr) [reply src/3rd_party/spdlog/async_logger.h:1]: if not, I may need to re-add them later

File: CMakeLists.txt
Comment (vrnmthr) [reply CMakeLists.txt:287]: @snukky @hieuhoang would you prefer that I leave the -march in or take it out?

File: src/functional/operators.h
Comment (vrnmthr) [reply src/functional/operators.h:223]: we require ARMv8 which has NEON in cmake comment

File: CMakeLists.txt
Comment (vrnmthr) [reply CMakeLists.txt:108]: automatically detecting is a bit tough but I agreee is a good goal. I've made it clear that for now we only support v8 and will work on adding auto archdetect in future so as to not block this PR

File: CMakeLists.txt
Comment (vrnmthr) [reply CMakeLists.txt:287]: I am going to leave it as is right now to not break any backwardscompat. if we want, we can take it out later

File: .github/workflows/ios.yml
Comment (vrnmthr) [reply .github/workflows/ios.yml:1]: This cross compiles for iOS because it builds with the iOS toolchain. That means that it builds an ARM binary. I don't think we should combine them since macos pipeline builds an Intel binary

File: src/3rd_party/sentencepiece
Comment (emjotde) [inline src/3rd_party/sentencepiece:1]: We are updating sentencepiece? Was there a corresponding change in the sentencepiece repo?

File: src/3rd_party/faiss/VectorTransform.cpp
Comment (emjotde) [inline src/3rd_party/faiss/VectorTransform.cpp:23]: Add a comment why this is here.

File: src/common/binary.cpp
Comment (emjotde) [inline src/common/binary.cpp:112]: How sure are we this is always OK?

**Review by emjotde — COMMENTED** [review]
Request for small clarifications, will be good to merge afterwards.

File: src/3rd_party/sentencepiece
Comment (vrnmthr) [reply src/3rd_party/sentencepiece:1]: There was -- I made some minor changes to the cmake in sentencepiece to build for iOS correctly

Comment (vrnmthr) [reply src/3rd_party/sentencepiece:1]: I think github will allow you to see the diff in submodule if you click on the diff here

File: src/common/binary.cpp
Comment (vrnmthr) [reply src/common/binary.cpp:112]: I have tested this with the iOS build and with through our quicksand paths and have had no issues. If there are other scenarios that you know of that you want to test to ensure they don't break we can try and test them as well

Comment (vrnmthr) [reply src/common/binary.cpp:112]: in binary.cpp in the loadItems function there is a comment that indicates that perhaps there may be some hardware-specific compatibility we should test?
// @TOOD: verify this actually works for the hardware-specific ones like intgemm8avx2

File: src/3rd_party/faiss/VectorTransform.cpp
Comment (vrnmthr) [reply src/3rd_party/faiss/VectorTransform.cpp:23]: added

File: src/common/binary.cpp
Comment (vrnmthr) [reply src/common/binary.cpp:112]: if you are uncomfortable with this, we can also pass the mapped variable through the callchain to let callers set it as they see fit. we need this change in iOS otherwise loading/parsing the YAML is too slow

**Review by emjotde — APPROVED** [review]

File: src/common/binary.cpp
Comment (emjotde) [reply src/common/binary.cpp:112]: There have been a lot of changes in the mmapping interface in internal master, so we will take a look in the merge PR what is going on there.

**Pull request merged**
